src/player_activity.h

```diff
@@ -4,68 +4,79 @@
 #include "enums.h"
 #include "json.h"
 #include "item_location.h"
+#include "string_id.h"
 
 #include <climits>
 #include <vector>
 
-enum activity_type : int {    // expanded this enum for readability
-    ACT_NULL = 0,
-    ACT_RELOAD,
-    ACT_READ,
-    ACT_GAME,
-    ACT_WAIT,
-    ACT_CRAFT,
-    ACT_LONGCRAFT,
-    ACT_DISASSEMBLE,
-    ACT_BUTCHER,
-    ACT_LONGSALVAGE,
-    ACT_FORAGE,
-    ACT_BUILD,
-    ACT_VEHICLE,
-    ACT_REFILL_VEHICLE, // not used anymore.
-    ACT_TRAIN,
-    ACT_WAIT_WEATHER,
-    ACT_FIRSTAID,
-    ACT_FISH,
-    ACT_PICKAXE,
-    ACT_BURROW,
-    ACT_PULP,
-    ACT_VIBE,
-    ACT_MAKE_ZLAVE,
-    ACT_DROP,
-    ACT_STASH,
-    ACT_PICKUP,
-    ACT_MOVE_ITEMS,
-    ACT_ADV_INVENTORY,
-    ACT_ARMOR_LAYERS,
-    ACT_START_FIRE,
-    ACT_OPEN_GATE,
-    ACT_FILL_LIQUID,
-    ACT_HOTWIRE_CAR,
-    ACT_AIM,
-    ACT_ATM,
-    ACT_START_ENGINES,
-    ACT_OXYTORCH,
-    ACT_CRACKING,
-    ACT_REPAIR_ITEM,
-    ACT_MEND_ITEM,
-    ACT_GUNMOD_ADD,
-    ACT_WAIT_NPC,
-    ACT_CLEAR_RUBBLE,
-    ACT_MEDITATE,
-    NUM_ACTIVITIES
-};
-
 class player;
 class Character;
+class player_activity;
+class activity_type;
+
+using activity_id = string_id<activity_type>;
+
+enum based_on_type : int {
+    TIME = 0,
+    SPEED,
+    NEITHER
+};
+
+/** A class that stores constant information that doesn't differ between activities of the same type */
+class activity_type
+{
+    private:
+        string_id<activity_type> id_;
+        bool rooted_;
+        std::string stop_phrase_;
+        bool abortable_;
+        bool suspendable_;
+        based_on_type based_on_;
+        static std::map< string_id<activity_type>, activity_type > all_;
+
+    public:
+        const string_id<activity_type> &id() const {
+            return id_;
+        }
+        bool rooted() const {
+            return rooted_;
+        }
+        bool abortable() const {
+            return abortable_;
+        }
+        bool suspendable() const {
+            return suspendable_;
+        }
+        const std::string &stop_phrase() const {
```

Comment:
It returns a reference to the private variable in `activity_type`. The `activity_type` objects shouldn't die until sometime after the player saves/suicides, and at that point the stop phrase shouldn't be used (so I should change the return to a `std::string` just in case it is used?).